File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.java
import java.io.*;
import java.util.*;

/**
 * O(n^2) or nmlogn까지
 * -> 전체를 스위핑 하고 끝나야 함. 이전의 상태를 바탕으로 -> 현재 결정: DP
 * d[i][j] : (i, j)가 꼭짓점으로 끝나는 정사각형의 길이
 * d[i][j] : 
 */
// 
public class Main {

    static char[][] board = new char[1002][1002];
    static int[][] d = new int[1002][1002];
    static int n;
    static int m;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] parts = br.readLine().split(" ");
        n = Integer.parseInt(parts[0]);
        m = Integer.parseInt(parts[1]);

        for (int i = 1; i <= n; i++) {
            String input = br.readLine();
            for (int j = 1; j <= m; j++) {
                board[i][j] = input.charAt(j - 1);
            }
        }
        int max = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (board[i][j] == '0') {
                    d[i][j] = 0;
                } else {
                    d[i][j] = Math.min(Math.min(d[i - 1][j - 1], d[i - 1][j]), d[i][j - 1]) + 1;
                }
                max = Math.max(d[i][j], max);
            }
        }
        System.out.print(max * max);
    }
}

File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.kt
/**
 * d[i][j]: i * j에서 가장 큰 정사각형 크기 -> 다른 말로 하면 가장 긴 한 변의 길이
 * - arr[i - 1][j] 도 1, arr[i][j - 1]도 1, arr[i - 1][j - 1]도 1
 */

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val arr = Array(n + 1) { IntArray(m + 1) }
    val d = Array(n + 1) { IntArray(m + 1) }
    for (i in 1..n) {
        val line = readLine()
        for (j in 1..m) {
            arr[i][j] = line[j - 1] - '0'
        }
    }

    for (i in 1..n) {
        for (j in 1..m) {
            if (arr[i][j] == 0) continue
            if (arr[i - 1][j] == 1 && arr[i][j - 1] == 1 && arr[i - 1][j - 1] == 1) {
                d[i][j] = minOf(d[i - 1][j], d[i][j - 1], d[i - 1][j - 1]) + 1
            } else {
                d[i][j] = 1
            }
        }
    }

    val length = d.maxOf { line -> line.max() }
    print(length * length)
}